### DISCUSSION TITLE
[FEATURE REQUEST] Support for repo format DEB822

### DISCUSSION THREAD

**Discussion by ppoudrier-omnimed** [saltstack/salt#67334] [Feature Requests]
**Is your feature request related to a problem? Please describe.**
Actually, when I use the following state, it always get executed.
```
suricata-repo:
  pkgrepo.managed:
    - ppa: oisf/suricata-stable
```
I digged deeper to find that the repo is never listed in `salt 'host' pkg.list_repos`
_Please note that le ppa of suricata for Ubuntu 24.04 always return the deb822 format and if I try to force the use of keyserver and keyid, the key is invalid to apt_

**Describe the solution you'd like**
I would like have a state that do not reinstall the repo each time so I can have a valid test=true to know if something need to be applied

**Describe alternatives you've considered**
I ended using file.managed

**Additional context**
- Tested with salt version 3006.9 (Master and Minion)
- Master on Ubuntu 20.04 and Minion on Ubuntu 24.04

**Please Note**
If this feature request would be considered a substantial change or addition, this should go through a SEP process here https://github.com/saltstack/salt-enhancement-proposals, instead of a feature request.

**Comment by welcome** [comment]
Hi there! Welcome to the Salt Community! Thank you for making your first contribution. We have a lengthy process for issues and PRs. Someone from the Core Team will follow up as soon as possible. In the meantime, here’s some information that may help as you continue your Salt journey.
Please be sure to review our [Code of Conduct](https://github.com/saltstack/salt/blob/master/CODE_OF_CONDUCT.md). Also, check out some of our community resources including:

  - [Community Wiki](https://github.com/saltstack/community/wiki)
  - [Salt’s Contributor Guide](https://docs.saltproject.io/en/master/topics/development/contributing.html)
  - [Join our Community Slack](https://via.vmw.com/salt-slack)
  - [IRC on LiberaChat](https://web.libera.chat/#salt)
  - [Salt Project YouTube channel](https://www.youtube.com/channel/UCpveTIucFx9ljGelW63-BWg)
  - [Salt Project Twitch channel](https://www.twitch.tv/saltprojectoss)

There are lots of ways to get involved in our community. Every month, there are around a dozen opportunities to meet with other contributors and the Salt Core team and collaborate in real time. The best way to keep track is by subscribing to the Salt Community Events Calendar.
If you have additional questions, email us at saltproject@vmware.com. We’re glad you’ve joined our community and look forward to doing awesome things with you!

**Comment by teutat3s** [comment]
With https://github.com/saltstack/salt-install-guide/pull/6 merged it is now no longer possible to install `salt-*` packages with salt `3006.9`. Following the [official installation docs](https://docs.saltproject.io/salt/install-guide/en/latest/topics/install-by-operating-system/linux-deb.html) for Debian, this error appears on a Debian 12 minion:

```
----------
          ID: salt-minion
    Function: pkg.installed
      Result: False
     Comment: An exception occurred in this state: Traceback (most recent call last):
                File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/state.py", line 2440, in call
                  ret = self.states[cdata["full"]](
                File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/loader/lazy.py", line 159, in __call__
                  ret = self.loader.run(run_func, *args, **kwargs)
                File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/loader/lazy.py", line 1245, in run
                  return self._last_context.run(self._run_as, _func_or_method, *args, **kwargs)
                File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/loader/lazy.py", line 1260, in _run_as
                  ret = _func_or_method(*args, **kwargs)
                File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/loader/lazy.py", line 1296, in wrapper
                  return f(*args, **kwargs)
                File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/states/pkg.py", line 1703, in installed
                  result = _find_install_targets(
                File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/states/pkg.py", line 716, in _find_install_targets
                  resolved_latest = __salt__["pkg.latest_version"](
                File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/loader/lazy.py", line 159, in __call__
                  ret = self.loader.run(run_func, *args, **kwargs)
                File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/loader/lazy.py", line 1245, in run
                  return self._last_context.run(self._run_as, _func_or_method, *args, **kwargs)
                File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/loader/lazy.py", line 1260, in _run_as
                  ret = _func_or_method(*args, **kwargs)
                File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/modules/aptpkg.py", line 484, in latest_version
                  refresh_db(cache_valid_time)
                File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/modules/aptpkg.py", line 609, in refresh_db
                  raise CommandExecutionError(comment)
              salt.exceptions.CommandExecutionError: E: Type 'X-Repolib-Name:' is not known on line 1 in source list /etc/apt/sources.list.d/salt.list
              E: The list of sources could not be read.
```

EDIT: Workaround for the curious reader: Write your own `salt.list`
```
deb [signed-by=/etc/apt/keyrings/salt-archive-keyring.gpg arch=amd64] https://packages.broadcom.com/artifactory/saltproject-deb stable main
```

**Reply by dwoz** [reply]
@ScriptAutomate thoughts?

**Reply by m-czernek** [reply]
So we're working on adding deb822 support in https://github.com/saltstack/salt/pull/67748

An issue I noticed with upstream Salt testing on Ubuntu is that the Ubuntu package uses a venv, which is isolated from system libraries. This venv is missing the `python3-apt` dependency. 

So apart from that PR, Salt might have to either add the `python3-apt` dependency into the venv, or allow the venv to take system libraries (which is what we do in the `venv-salt-minion` package from SUSE)

**Reply by whytewolf** [reply]
HAS_APT was removed in 3007 and onwords. relenv is more than a venv. it is a completely sideported python. this raises issues with including things like python3-apt. which can not be pip installed. 

the other reason for has_apt being removed is the split code functionality it caused. with half the code capable of things only if python-apt was available.

**Reply by dwoz** [reply]
> So apart from that PR, Salt might have to either add the `python3-apt` dependency into the venv, or allow the venv to take system libraries (which is what we do in the `venv-salt-minion` package from SUSE)

The problem we'll run into with pulling in system python libraries is potential variance in python versions. Things can break quickly if our packaged python version > or < the system python. I think a calling a wrapper around said modules would be a better approach. That said, in this case, due to the split in functionality @whytewolf mentioned I think we should lean into using the cli tools apt provides.

**Reply by m-czernek** [reply]
I'll take a look how this is done in 3007 and onwards. It's true that there is the `add-apt-repository` CLI tool. Then again, this seems to have issues with deb822 as well [1], at least for now. I'm also not sure whether it's missing some features for managing repositories, though it's definitely worth researching/experimenting with.

[1] https://bugs.launchpad.net/ubuntu/+source/software-properties/+bug/2052851

**Reply by m-czernek** [reply]
Folks, we've changed the approach to the feature in the following PR: https://github.com/saltstack/salt/pull/67956

Please take a look; this approach is consistent with Salt's deb-management approach on master, i.e. not using any system python libraries and relying on internal classes.
